sdk/Sdk/FunctionMetadataGenerator.cs

```diff
@@ -378,14 +372,233 @@ private static ExpandoObject BuildBindingMetadataFromAttribute(CustomAttribute a
+                && isBatchedValue is bool isBatched)
+            {
+                // Batching set to true
+                if (isBatched)
+                {
+                    bindingDict["Cardinality"] = "Many";
+                    // Throw if parameter type is *definitely* not a collection type.
+                    // Note that this logic doesn't dictate what we can/can't do, and
+                    // we can be more restrictive in the future because today some 
+                    // scenarios result in runtime failures.
+                    if (IsIterableCollection(parameterType, out DataType dataType))
+                    {
+                        if (dataType.Equals(DataType.String))
+                        {
+                            bindingDict["DataType"] = "String";
+                        }
+                        else if (dataType.Equals(DataType.Binary))
+                        {
+                            bindingDict["DataType"] = "Binary";
+                        }
+                    }
+                    else
+                    {
+                        throw new FunctionsMetadataGenerationException("Function is configured to process events in batches but parameter type is not iterable. " +
+                            $"Change parameter named '{ parameterName }' to be an IEnumerable type or set 'IsBatched' to false on your '{attribute.AttributeType.Name.Replace("Attribute", "")}' attribute.");
+                    }
+                }
+                // Batching set to false
+                else
+                {
+                    bindingDict["Cardinality"] = "One";
+                }
+
+                bindingDict.Remove(Constants.IsBatchedKey);
+            }
+
             return binding;
         }
 
+        private static bool IsIterableCollection(TypeReference type, out DataType dataType)
+        {
+            // Array and not byte array 
+            bool isArray = type.IsArray && !string.Equals(type.FullName, Constants.ByteArrayType, StringComparison.Ordinal);
+            if (isArray)
+            {
+                TypeSpecification? typeSpecification = type as TypeSpecification;
+                if (typeSpecification is not null)
+                {
+                    dataType = GetDataTypeFromType(typeSpecification.ElementType.FullName);
+                    return true;
+                }
+            }
+
+            bool isMappingEnumerable = IsOrDerivedFrom(type, Constants.IEnumerableOfKeyValuePair)
+                || IsOrDerivedFrom(type, Constants.LookupGenericType)
+                || IsOrDerivedFrom(type, Constants.DictionaryGenericType);
+            if (isMappingEnumerable)
+            {
+                dataType = DataType.Undefined;
+                return false;
+            }
+
+            // IEnumerable and not string or dictionary
+            bool isEnumerableOfT = IsOrDerivedFrom(type, Constants.IEnumerableOfT);
```

Comment:
nit: Since this may not be used until you reach the last check, could we defer this check?

Reply:
is there a fancy .net thing i can do? I use `isEnumerableOfT` a bit earlier in line 473 too